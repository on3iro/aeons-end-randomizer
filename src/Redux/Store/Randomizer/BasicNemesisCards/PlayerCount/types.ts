import { ActionsUnion } from '@martin_hotell/rex-tils'

import * as types from 'aer-types/types'

import { actions } from './actions'

export type State = PlayerCount

export enum ActionTypes {
  SET = 'BasicNemesisCards/PlayerCount/SET',
}

export type Action = ActionsUnion<typeof actions>

export type PlayerCount = types.PlayerCount

export type PlayerCountStateSlice = {
  Randomizer: {
    BasicNemesisCards: {
      PlayerCount: PlayerCount
    }
  }
}
