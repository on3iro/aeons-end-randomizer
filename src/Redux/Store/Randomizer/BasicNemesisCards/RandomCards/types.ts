import { ActionsUnion } from '@martin_hotell/rex-tils'

import { actions } from './actions'

export type State = Readonly<{
  Tier1: ReadonlyArray<Readonly<{ id: string }>>
  Tier2: ReadonlyArray<Readonly<{ id: string }>>
  Tier3: ReadonlyArray<Readonly<{ id: string }>>
  recoverySeed?: string
}>

export enum ActionTypes {
  CREATE = 'BasicNemesisCards/RandomCards/CREATE',
}

export type Action = ActionsUnion<typeof actions>

export type RandomCardsStateSlice = {
  Randomizer: {
    BasicNemesisCards: {
      RandomCards: State
    }
  }
}
