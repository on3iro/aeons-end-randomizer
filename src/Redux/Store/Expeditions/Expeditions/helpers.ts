import shortid from 'shortid'

import * as types from 'types'
import { byCost } from 'helpers'

import {
  createArrayWithDefaultValues,
  createIdList,
  getRandomEntity,
  createSupply,
} from 'Redux/helpers'

import { BaseConfig, RollBattleConfig, WinConfig, LossConfig } from './types'

export const calcBattleScore = (tries: number) => {
  switch (tries) {
    case 1: {
      return 6
    }

    case 2: {
      return 4
    }

    case 3: {
      return 2
    }

    default: {
      return 0
    }
  }
}

export const determineBaseTier = () => {}

export const generateBattles = (
  variant: types.Variant,
  expeditionId: string
) => {
  const battles = variant.configList.map(
    (config, index): types.Battle => {
      const isFirst = index === 0

      return {
        id: shortid.generate(),
        nemesisTier: config.tier,
        treasure: config.treasure,
        expeditionId,
        status: isFirst ? 'unlocked' : 'locked',
        tries: 0,
      }
    }
  )

  return battles
}

export const createExpeditionConfig = ({
  variant,
  name,
  bigPocketVariant,
  availableMageIds,
  availableCards,
  availableLevel1TreasureIds,
  tiles,
}: BaseConfig) => {
  const mageIds = createIdList(
    availableMageIds,
    createArrayWithDefaultValues(4, 'EMPTY'),
    getRandomEntity
  ).result
  const { gems, relics, spells } = createSupply(availableCards, tiles)
  const gemsByCost = gems.result.sort(byCost)
  const relicsByCost = relics.result.sort(byCost)
  const spellsByCost = spells.result.sort(byCost)

  const supplyIds = [...gemsByCost, ...relicsByCost, ...spellsByCost].map(
    card => card.id
  )

  const startsWithTreasure = variant.configList[0].tier.tier > 1 // TODO should probably use our rollTreasure property
  const treasureIds = startsWithTreasure
    ? createIdList(
        availableLevel1TreasureIds,
        createArrayWithDefaultValues(5, 'EMPTY'),
        getRandomEntity
      ).result
    : []

  const expeditionId = shortid.generate()
  const battles = generateBattles(variant, expeditionId)

  return {
    battles,
    bigPocketVariant,
    expeditionId,
    mageIds,
    name,
    supplyIds,
    treasureIds,
    variantId: variant.id,
  }
}

// Because we always add these cards to an existing array inside our expedition in our store,
// the count of newly added cards decreases by tier.
// If we ever have a variant, that starts higher than tier 2 we have to
// incorporate changes to roll the whole amount of cards per tier!
const getUpgradedBasicNemesisIdsByBattleTier = ({
  availableTier1Ids,
  availableTier2Ids,
  availableTier3Ids,
  battleTier,
}: {
  availableTier1Ids: string[]
  availableTier2Ids: string[]
  availableTier3Ids: string[]
  battleTier: 1 | 2 | 3 | 4
}) => {
  switch (battleTier) {
    case 1: {
      return [] // No upgraded cards are added on tier 1
    }

    case 2: {
      const tier1Ids = createIdList(
        availableTier1Ids,
        createArrayWithDefaultValues(1, 'EMPTY'),
        getRandomEntity
      ).result
      const tier2Ids = createIdList(
        availableTier2Ids,
        createArrayWithDefaultValues(3, 'EMPTY'),
        getRandomEntity
      ).result
      const tier3Ids = createIdList(
        availableTier3Ids,
        createArrayWithDefaultValues(3, 'EMPTY'),
        getRandomEntity
      ).result

      return [...tier1Ids, ...tier2Ids, ...tier3Ids]
    }

    case 3:
    case 4: {
      const tier1Ids = createIdList(
        availableTier1Ids,
        createArrayWithDefaultValues(1, 'EMPTY'),
        getRandomEntity
      ).result
      const tier2Ids = createIdList(
        availableTier2Ids,
        createArrayWithDefaultValues(1, 'EMPTY'),
        getRandomEntity
      ).result
      const tier3Ids = createIdList(
        availableTier3Ids,
        createArrayWithDefaultValues(2, 'EMPTY'),
        getRandomEntity
      ).result

      return [...tier1Ids, ...tier2Ids, ...tier3Ids]
    }

    default: {
      return [] // Should never occur!
    }
  }
}

const rollNewUpgradedNemesisCards = (
  availableUpgradedBasicNemesisCards: types.UpgradedBasicNemesisCard[],
  previousUpgradedBasicNemesisCards: string[],
  nemesisTier: 1 | 2 | 3 | 4
) => {
  const upgradedCardsWithoutPreviousCards = availableUpgradedBasicNemesisCards.filter(
    upgradedCard => !previousUpgradedBasicNemesisCards.includes(upgradedCard.id)
  )

  const tier1AvailableUpgradedNemesisIds = upgradedCardsWithoutPreviousCards
    .filter(card => card.tier === 1)
    .map(card => card.id)
  const tier2AvailableUpgradedNemesisIds = upgradedCardsWithoutPreviousCards
    .filter(card => card.tier === 2)
    .map(card => card.id)
  const tier3AvailableUpgradedNemesisIds = upgradedCardsWithoutPreviousCards
    .filter(card => card.tier === 3)
    .map(card => card.id)

  const upgradedBasicNemesisCardIds = getUpgradedBasicNemesisIdsByBattleTier({
    battleTier: nemesisTier,
    availableTier1Ids: tier1AvailableUpgradedNemesisIds,
    availableTier2Ids: tier2AvailableUpgradedNemesisIds,
    availableTier3Ids: tier3AvailableUpgradedNemesisIds,
  })

  return [...previousUpgradedBasicNemesisCards, ...upgradedBasicNemesisCardIds]
}

export const createBattle = (config: RollBattleConfig) => {
  const nemesisIds = config.availableNemeses
    .filter(
      nemesis => nemesis.expeditionRating === config.battle.nemesisTier.tier
    )
    .map(nemesis => nemesis.id)
    .filter(nemesisId => !config.previousNemeses.includes(nemesisId))

  const nemesisId = createIdList(
    nemesisIds,
    createArrayWithDefaultValues(1, 'EMPTY'),
    getRandomEntity
  ).result[0]

  const upgradedBasicNemesisCardIds = config.battle.nemesisTier.isNewTier
    ? rollNewUpgradedNemesisCards(
        config.availableUpgradedBasicNemesisCards,
        config.previousUpgradedBasicNemesisCards,
        config.battle.nemesisTier.tier
      )
    : config.previousUpgradedBasicNemesisCards

  return {
    battle: { ...config.battle, nemesisId, status: 'before_battle' },
    upgradedBasicNemesisCardIds,
  }
}

const rollNewEntity = (list: string[]): string =>
  createIdList(list, createArrayWithDefaultValues(1, 'EMPTY'), getRandomEntity)
    .result[0]

export const rollWinRewards = (config: WinConfig) => {
  const newTreasures = createIdList(
    config.treasureIds,
    createArrayWithDefaultValues(5, 'EMPTY'),
    getRandomEntity
  ).result

  const newGem = rollNewEntity(config.gemIds)
  const newRelic = rollNewEntity(config.relicIds)
  const newSpell = rollNewEntity(config.spellIds)

  const updatedBattle = {
    ...config.battle,
    rewards: {
      treasure: newTreasures,
      supplyIds: [newGem, newRelic, newSpell],
      mage: undefined, // we explicitely overwrite the mage reward from earlier losses
    },
  }
  return { battle: updatedBattle }
}

export const rollLossRewards = (config: LossConfig) => {
  switch (config.type) {
    case 'mage': {
      const newMage = rollNewEntity(config.mageIds)

      return {
        ...config.battle,
        rewards: {
          treasure: [],
          mage: newMage,
          supplyIds: [],
        },
      }
    }

    case 'gem': {
      const newGem = rollNewEntity(config.gemIds)

      return {
        ...config.battle,
        rewards: {
          treasure: [],
          mage: undefined,
          supplyIds: [newGem],
        },
      }
    }

    case 'relic': {
      const newRelic = rollNewEntity(config.relicIds)

      return {
        ...config.battle,
        rewards: {
          treasure: [],
          mage: undefined,
          supplyIds: [newRelic],
        },
      }
    }

    case 'spell': {
      const newSpell = rollNewEntity(config.spellIds)

      return {
        ...config.battle,
        rewards: {
          treasure: [],
          mage: undefined,
          supplyIds: [newSpell],
        },
      }
    }

    case 'treasure1': {
      const newTreasure = rollNewEntity(config.treasure1Ids)

      return {
        ...config.battle,
        rewards: {
          treasure: [newTreasure],
          mage: undefined,
          supplyIds: [],
        },
      }
    }

    case 'treasure2': {
      const newTreasure = rollNewEntity(config.treasure2Ids)

      return {
        ...config.battle,
        rewards: {
          treasure: [newTreasure],
          mage: undefined,
          supplyIds: [],
        },
      }
    }

    case 'treasure3': {
      const newTreasure = rollNewEntity(config.treasure3Ids)

      return {
        ...config.battle,
        rewards: {
          treasure: [newTreasure],
          mage: undefined,
          supplyIds: [],
        },
      }
    }

    default: {
      return { ...config.battle }
    }
  }
}
