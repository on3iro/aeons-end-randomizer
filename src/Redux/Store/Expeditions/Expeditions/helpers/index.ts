export * from './convertExpeditionToConfig'
export * from './convertExpeditionFromConfig'

export const calcBattleScore = (tries: number) => {
  switch (tries) {
    case 1: {
      return 6
    }

    case 2: {
      return 4
    }

    case 3: {
      return 2
    }

    default: {
      return 0
    }
  }
}
