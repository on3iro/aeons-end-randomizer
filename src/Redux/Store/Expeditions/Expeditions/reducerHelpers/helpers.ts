import { loop, Cmd } from 'redux-loop'
import { set as setToDb } from 'idb-keyval'
import * as types from 'aer-types/types'

import { State } from '../types'
import { actions } from '../actions'

export const EXPEDITIONS_DB_KEY = 'expeditions-1.11.0'

export const updateBattle = (
  state: State,
  battle: types.Battle,
  additionalBattleProps?: {
    status?: types.BattleStatus
    tries?: number
    rewards?: types.Rewards
  }
) => {
  const oldExpedition = state.expeditions[battle.expeditionId]
  const branches = oldExpedition.sequence.branches

  const updatedBranches = {
    ...branches,
    [battle.id]: {
      ...battle,
      ...additionalBattleProps,
    },
  }

  const newState = {
    ...state,
    expeditions: {
      ...state.expeditions,
      [battle.expeditionId]: {
        ...oldExpedition,
        sequence: {
          ...oldExpedition.sequence,
          branches: updatedBranches,
        },
      },
    },
  }

  return loop(
    newState,
    Cmd.run(setToDb, {
      args: [EXPEDITIONS_DB_KEY, newState],
      successActionCreator: actions.setToDBSuccessful,
      failActionCreator: actions.setToDBFailed,
    })
  )
}
