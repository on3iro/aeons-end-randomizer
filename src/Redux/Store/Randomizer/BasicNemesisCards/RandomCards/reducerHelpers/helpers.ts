import seedrandom from 'seedrandom'
import shortid from 'shortid'

import * as types from 'aer-types/types'

import {
  getRandomEntity,
  createSlotList,
  createBasicNemesisCardList,
} from 'Redux/helpers'

type CardMatrix = {
  [key: string]: {
    [key: number]: number
  }
}

export const cardMatrix: CardMatrix = {
  tier1: {
    1: 1,
    2: 3,
    3: 5,
    4: 8,
  },
  tier2: {
    1: 3,
    2: 5,
    3: 6,
    4: 7,
  },
  tier3: {
    1: 7,
    2: 7,
    3: 7,
    4: 7,
  },
}

export const getBasicNemesisCardsByTier = (
  cards: ReadonlyArray<types.BasicNemesisCard>,
  tier: types.NemesisCardTier
) => cards.filter((card) => card.tier === tier)

export const getRandomMinionAmount = (
  minCount: number = 1,
  count: number = 3,
  seed: types.Seed = {
    seed: shortid.generate(),
  }
) => {
  const rng = seedrandom(seed.seed)
  return (
    Math.floor(rng() * (Math.floor(count) - Math.ceil(minCount))) +
    Math.ceil(minCount)
  )
}

export const drawMinions = (
  cards: ReadonlyArray<types.BasicNemesisCard>,
  amount: number,
  seed?: types.Seed
) => {
  const minions = cards.filter((card) => card.type === 'Minion')
  const minionsList = createBasicNemesisCardList(
    minions,
    createSlotList(amount),
    getRandomEntity,
    seed
  ).result

  return minionsList.map((minion) => {
    return {
      id: minion.id,
    }
  })
}

export const drawPowersAndAttacks = (
  cards: ReadonlyArray<types.BasicNemesisCard>,
  amount: number,
  seed?: types.Seed
) => {
  const powersAndAttacks = cards.filter((card) => card.type !== 'Minion')
  const powersAndAttacksList = createBasicNemesisCardList(
    powersAndAttacks,
    createSlotList(amount),
    getRandomEntity,
    seed
  ).result

  return powersAndAttacksList.map((powerAndAttack) => {
    return {
      id: powerAndAttack.id,
    }
  })
}

export const getRandomBasicNemesisCardsByTier = (
  cards: ReadonlyArray<types.BasicNemesisCard>,
  tier: types.NemesisCardTier,
  playerCount: types.PlayerCount,
  seed?: types.Seed
) => {
  // Get cards by tier
  const cardsByTier = getBasicNemesisCardsByTier(cards, tier)

  // Get amount of minions per player
  const minionAmount =
    playerCount === 1 && tier === 1
      ? getRandomMinionAmount(0, 2, seed)
      : getRandomMinionAmount(1, 3, seed)

  // Draw up to max amount of minions from minions
  const randomMinions = drawMinions(cardsByTier, minionAmount, seed)

  // Draw tierCount - minionCount cards from power/attack pool
  const randomPowersAndAttacks = drawPowersAndAttacks(
    cardsByTier,
    cardMatrix[`tier${tier}`][playerCount] - randomMinions.length,
    seed
  )

  return [...randomMinions, ...randomPowersAndAttacks]
}

export const getRandomBasicNemesisCardList = (
  cards: ReadonlyArray<types.BasicNemesisCard>,
  playerCount: types.PlayerCount,
  seed?: types.Seed
) => {
  return {
    Tier1: getRandomBasicNemesisCardsByTier(cards, 1, playerCount, seed),
    Tier2: getRandomBasicNemesisCardsByTier(cards, 2, playerCount, seed),
    Tier3: getRandomBasicNemesisCardsByTier(cards, 3, playerCount, seed),
  }
}
