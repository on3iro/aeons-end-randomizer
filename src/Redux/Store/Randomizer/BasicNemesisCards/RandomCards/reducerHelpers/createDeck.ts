import { State } from '../types'

import { actions } from '../actions'

import { getRandomBasicNemesisCardList } from './helpers'

export const createDeck = (
  state: State,
  action: ReturnType<typeof actions.createBasicNemesisCardDeck>
) => {
  const { availableCards, playerCount, seed } = action.payload

  const newState = getRandomBasicNemesisCardList(
    availableCards,
    playerCount,
    seed
  )

  return {
    ...state,
    Tier1: newState.Tier1,
    Tier2: newState.Tier2,
    Tier3: newState.Tier3,
  }
}
