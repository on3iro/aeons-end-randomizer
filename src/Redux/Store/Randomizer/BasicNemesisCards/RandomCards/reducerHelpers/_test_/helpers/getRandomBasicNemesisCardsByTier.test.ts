import * as types from 'types'

import { getRandomBasicNemesisCardsByTier } from 'Redux/Store/Randomizer/BasicNemesisCards/RandomCards/reducerHelpers/helpers'

const cards = [
  { id: 'minion1', type: 'Minion', tier: 1 },
  { id: 'minion2', type: 'Minion', tier: 1 },
  { id: 'minion3', type: 'Minion', tier: 1 },
  { id: 'minion4', type: 'Minion', tier: 2 },
  { id: 'minion5', type: 'Minion', tier: 2 },
  { id: 'minion6', type: 'Minion', tier: 2 },
  { id: 'minion7', type: 'Minion', tier: 3 },
  { id: 'minion8', type: 'Minion', tier: 3 },
  { id: 'minion9', type: 'Minion', tier: 3 },
  { id: 'power1', type: 'Power', tier: 1 },
  { id: 'power2', type: 'Power', tier: 1 },
  { id: 'power3', type: 'Power', tier: 1 },
  { id: 'power4', type: 'Power', tier: 2 },
  { id: 'power5', type: 'Power', tier: 2 },
  { id: 'power6', type: 'Power', tier: 2 },
  { id: 'power7', type: 'Power', tier: 3 },
  { id: 'power8', type: 'Power', tier: 3 },
  { id: 'power9', type: 'Power', tier: 3 },
  { id: 'attack1', type: 'Attack', tier: 1 },
  { id: 'attack2', type: 'Attack', tier: 1 },
  { id: 'attack3', type: 'Attack', tier: 1 },
  { id: 'attack4', type: 'Attack', tier: 2 },
  { id: 'attack5', type: 'Attack', tier: 2 },
  { id: 'attack6', type: 'Attack', tier: 2 },
  { id: 'attack7', type: 'Attack', tier: 3 },
  { id: 'attack8', type: 'Attack', tier: 3 },
  { id: 'attack9', type: 'Attack', tier: 3 },
] as types.BasicNemesisCard[]

describe('getRandomBasicNemesisCardsByTier()', () => {
  it('should handle one player tier 1', () => {
    const result = getRandomBasicNemesisCardsByTier(cards, 1, 1, {
      seed: 'test',
    })
    expect(result).toMatchSnapshot()
    expect(result.length).toBe(1)
  })

  it('should handle one player tier 2', () => {
    const result = getRandomBasicNemesisCardsByTier(cards, 2, 1, {
      seed: 'test',
    })
    expect(result).toMatchSnapshot()
    expect(result.length).toBe(3)
  })

  it('should handle one player tier 3', () => {
    const result = getRandomBasicNemesisCardsByTier(cards, 3, 1, {
      seed: 'test',
    })
    expect(result).toMatchSnapshot()
    expect(result.length).toBe(7)
  })

  it('should handle two players tier 1', () => {
    const result = getRandomBasicNemesisCardsByTier(cards, 1, 2, {
      seed: 'test',
    })
    expect(result).toMatchSnapshot()
    expect(result.length).toBe(3)
  })

  it('should handle two players tier 2', () => {
    const result = getRandomBasicNemesisCardsByTier(cards, 2, 2, {
      seed: 'test',
    })
    expect(result).toMatchSnapshot()
    expect(result.length).toBe(5)
  })

  it('should handle two players tier 3', () => {
    const result = getRandomBasicNemesisCardsByTier(cards, 3, 2, {
      seed: 'test',
    })
    expect(result).toMatchSnapshot()
    expect(result.length).toBe(7)
  })

  it('should handle three players tier 1', () => {
    const result = getRandomBasicNemesisCardsByTier(cards, 1, 3, {
      seed: 'test',
    })
    expect(result).toMatchSnapshot()
    expect(result.length).toBe(5)
  })

  it('should handle three players tier 2', () => {
    const result = getRandomBasicNemesisCardsByTier(cards, 2, 3, {
      seed: 'test',
    })
    expect(result).toMatchSnapshot()
    expect(result.length).toBe(6)
  })

  it('should handle three players tier 3', () => {
    const result = getRandomBasicNemesisCardsByTier(cards, 3, 3, {
      seed: 'test',
    })
    expect(result).toMatchSnapshot()
    expect(result.length).toBe(7)
  })

  it('should handle four players tier 1', () => {
    const result = getRandomBasicNemesisCardsByTier(cards, 1, 4, {
      seed: 'test',
    })
    expect(result).toMatchSnapshot()
    expect(result.length).toBe(8)
  })

  it('should handle four players tier 2', () => {
    const result = getRandomBasicNemesisCardsByTier(cards, 2, 4, {
      seed: 'test',
    })
    expect(result).toMatchSnapshot()
    expect(result.length).toBe(7)
  })

  it('should handle four players tier 3', () => {
    const result = getRandomBasicNemesisCardsByTier(cards, 3, 4, {
      seed: 'test',
    })
    expect(result).toMatchSnapshot()
    expect(result.length).toBe(7)
  })
})
