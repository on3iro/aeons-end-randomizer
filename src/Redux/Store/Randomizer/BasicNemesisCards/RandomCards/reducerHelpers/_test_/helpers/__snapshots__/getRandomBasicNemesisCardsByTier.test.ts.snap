// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getRandomBasicNemesisCardsByTier() should handle four players tier 1 1`] = `
Array [
  Object {
    "id": "minion3",
    "tier": 1,
    "type": "Minion",
  },
  Object {
    "id": "minion1",
    "tier": 1,
    "type": "Minion",
  },
  Object {
    "id": "attack3",
    "tier": 1,
    "type": "Attack",
  },
  Object {
    "id": "power3",
    "tier": 1,
    "type": "Power",
  },
  Object {
    "id": "attack2",
    "tier": 1,
    "type": "Attack",
  },
  Object {
    "id": "power1",
    "tier": 1,
    "type": "Power",
  },
  Object {
    "id": "power2",
    "tier": 1,
    "type": "Power",
  },
  Object {
    "id": "attack1",
    "tier": 1,
    "type": "Attack",
  },
]
`;

exports[`getRandomBasicNemesisCardsByTier() should handle four players tier 2 1`] = `
Array [
  Object {
    "id": "minion6",
    "tier": 2,
    "type": "Minion",
  },
  Object {
    "id": "minion4",
    "tier": 2,
    "type": "Minion",
  },
  Object {
    "id": "attack6",
    "tier": 2,
    "type": "Attack",
  },
  Object {
    "id": "power6",
    "tier": 2,
    "type": "Power",
  },
  Object {
    "id": "attack5",
    "tier": 2,
    "type": "Attack",
  },
  Object {
    "id": "power4",
    "tier": 2,
    "type": "Power",
  },
  Object {
    "id": "power5",
    "tier": 2,
    "type": "Power",
  },
]
`;

exports[`getRandomBasicNemesisCardsByTier() should handle four players tier 3 1`] = `
Array [
  Object {
    "id": "minion9",
    "tier": 3,
    "type": "Minion",
  },
  Object {
    "id": "minion7",
    "tier": 3,
    "type": "Minion",
  },
  Object {
    "id": "attack9",
    "tier": 3,
    "type": "Attack",
  },
  Object {
    "id": "power9",
    "tier": 3,
    "type": "Power",
  },
  Object {
    "id": "attack8",
    "tier": 3,
    "type": "Attack",
  },
  Object {
    "id": "power7",
    "tier": 3,
    "type": "Power",
  },
  Object {
    "id": "power8",
    "tier": 3,
    "type": "Power",
  },
]
`;

exports[`getRandomBasicNemesisCardsByTier() should handle one player tier 1 1`] = `
Array [
  Object {
    "id": "minion3",
    "tier": 1,
    "type": "Minion",
  },
]
`;

exports[`getRandomBasicNemesisCardsByTier() should handle one player tier 2 1`] = `
Array [
  Object {
    "id": "minion6",
    "tier": 2,
    "type": "Minion",
  },
  Object {
    "id": "attack6",
    "tier": 2,
    "type": "Attack",
  },
  Object {
    "id": "power6",
    "tier": 2,
    "type": "Power",
  },
]
`;

exports[`getRandomBasicNemesisCardsByTier() should handle one player tier 3 1`] = `
Array [
  Object {
    "id": "minion9",
    "tier": 3,
    "type": "Minion",
  },
  Object {
    "id": "attack9",
    "tier": 3,
    "type": "Attack",
  },
  Object {
    "id": "power9",
    "tier": 3,
    "type": "Power",
  },
  Object {
    "id": "attack8",
    "tier": 3,
    "type": "Attack",
  },
  Object {
    "id": "power7",
    "tier": 3,
    "type": "Power",
  },
  Object {
    "id": "power8",
    "tier": 3,
    "type": "Power",
  },
  Object {
    "id": "attack7",
    "tier": 3,
    "type": "Attack",
  },
]
`;

exports[`getRandomBasicNemesisCardsByTier() should handle three players tier 1 1`] = `
Array [
  Object {
    "id": "minion3",
    "tier": 1,
    "type": "Minion",
  },
  Object {
    "id": "minion1",
    "tier": 1,
    "type": "Minion",
  },
  Object {
    "id": "attack3",
    "tier": 1,
    "type": "Attack",
  },
  Object {
    "id": "power3",
    "tier": 1,
    "type": "Power",
  },
  Object {
    "id": "attack2",
    "tier": 1,
    "type": "Attack",
  },
]
`;

exports[`getRandomBasicNemesisCardsByTier() should handle three players tier 2 1`] = `
Array [
  Object {
    "id": "minion6",
    "tier": 2,
    "type": "Minion",
  },
  Object {
    "id": "minion4",
    "tier": 2,
    "type": "Minion",
  },
  Object {
    "id": "attack6",
    "tier": 2,
    "type": "Attack",
  },
  Object {
    "id": "power6",
    "tier": 2,
    "type": "Power",
  },
  Object {
    "id": "attack5",
    "tier": 2,
    "type": "Attack",
  },
  Object {
    "id": "power4",
    "tier": 2,
    "type": "Power",
  },
]
`;

exports[`getRandomBasicNemesisCardsByTier() should handle three players tier 3 1`] = `
Array [
  Object {
    "id": "minion9",
    "tier": 3,
    "type": "Minion",
  },
  Object {
    "id": "minion7",
    "tier": 3,
    "type": "Minion",
  },
  Object {
    "id": "attack9",
    "tier": 3,
    "type": "Attack",
  },
  Object {
    "id": "power9",
    "tier": 3,
    "type": "Power",
  },
  Object {
    "id": "attack8",
    "tier": 3,
    "type": "Attack",
  },
  Object {
    "id": "power7",
    "tier": 3,
    "type": "Power",
  },
  Object {
    "id": "power8",
    "tier": 3,
    "type": "Power",
  },
]
`;

exports[`getRandomBasicNemesisCardsByTier() should handle two players tier 1 1`] = `
Array [
  Object {
    "id": "minion3",
    "tier": 1,
    "type": "Minion",
  },
  Object {
    "id": "minion1",
    "tier": 1,
    "type": "Minion",
  },
  Object {
    "id": "attack3",
    "tier": 1,
    "type": "Attack",
  },
]
`;

exports[`getRandomBasicNemesisCardsByTier() should handle two players tier 2 1`] = `
Array [
  Object {
    "id": "minion6",
    "tier": 2,
    "type": "Minion",
  },
  Object {
    "id": "minion4",
    "tier": 2,
    "type": "Minion",
  },
  Object {
    "id": "attack6",
    "tier": 2,
    "type": "Attack",
  },
  Object {
    "id": "power6",
    "tier": 2,
    "type": "Power",
  },
  Object {
    "id": "attack5",
    "tier": 2,
    "type": "Attack",
  },
]
`;

exports[`getRandomBasicNemesisCardsByTier() should handle two players tier 3 1`] = `
Array [
  Object {
    "id": "minion9",
    "tier": 3,
    "type": "Minion",
  },
  Object {
    "id": "minion7",
    "tier": 3,
    "type": "Minion",
  },
  Object {
    "id": "attack9",
    "tier": 3,
    "type": "Attack",
  },
  Object {
    "id": "power9",
    "tier": 3,
    "type": "Power",
  },
  Object {
    "id": "attack8",
    "tier": 3,
    "type": "Attack",
  },
  Object {
    "id": "power7",
    "tier": 3,
    "type": "Power",
  },
  Object {
    "id": "power8",
    "tier": 3,
    "type": "Power",
  },
]
`;
